cmd/download.go

```diff
@@ -110,3 +148,84 @@ func init() {
 	// is called directly, e.g.:
 	// downloadCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
 }
+
+// parseRangeArg parses episodes number with the following format
+// 1,2,3-5,!4 returns [1, 2, 3, 5]
+func parseRangeArg(arg string) ([]int, error) {
+	if len(arg) == 0 {
+		return nil, nil
+	}
+
+	// we try to convert the arg as a single episode
+	n, err := strconv.Atoi(arg)
+	if err == nil {
+		return []int{n}, nil
+	}
+
+	var results []int
+
+	// extract negative numbers !X
+	negatives := regexp.MustCompile(`!\d+`)
+	notWanted := negatives.FindAllString(arg, -1)
+
+	arg = negatives.ReplaceAllString(arg, "")
+
+	// extract ranges X-Y
+	rangesPattern := regexp.MustCompile(`\d+-\d+`)
+	ranges := rangesPattern.FindAllString(arg, -1)
+
+	arg = rangesPattern.ReplaceAllString(arg, "")
+
+	// extract the remaining single digit X
+	digitsPattern := regexp.MustCompile(`\d+`)
+	digits := digitsPattern.FindAllString(arg, -1)
+
+	for _, r := range ranges {
+		parts := strings.Split(r, "-")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("range %s must have the format start-end", r)
+		}
+
+		start, err := strconv.Atoi(parts[0])
+		if err != nil {
+			return nil, err
+		}
+
+		end, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, err
+		}
+
+		for i := start; i <= end; i++ {
+			// make sure it's wanted
+			wanted := true
+			for _, nw := range notWanted {
+				if fmt.Sprintf("!%d", i) == nw {
+					wanted = false
+					break
+				}
+			}
+
+			if !wanted {
+				continue
+			}
+
+			results = append(results, i)
+		}
+	}
+
+	// let's add the remaining digits
+	for _, d := range digits {
+		i, err := strconv.Atoi(d)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, i)
+	}
+
+	// we sort the result
+	sort.Ints(results)
+
```

Comment:
I think we probably have to do something to make this slice unique. For example, if someone types: `"1-5,!3,4-6"` that would currently cause to return [1, 2, 4, 4, 5, 5, 6].

So we can either create something to return a unique slice, or in the loop at line 183 already prevent doubles being inserted by using some set like structure (thinking of map[int]struct{}).

As it stands, it would not really have any side effects except an error being thrown for a file already existing/file system request fault. But I think it's an easy enough fix, right? 

What do you think? Worth it?

Reply:
yes good catch, I've added a map[int]bool that should handle this nicely.